### PR TITLE
emit error when doc generation fails

### LINKED ISSUE #41813: Compiler panics when generating documentation on full disk
When running cargo doc, rustc panics if disk is full. Not a huge issue, just an invalid call of `.unwrap()` and since the message says it's a bug, I'm filling a report.

Error message (with removed file path):
```
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'failed to generate documentation: Error { file: <path removed>", error: Error { repr: Os { code: 28, message: "No space left on device" } } }', /checkout/src/libcore/result.rs:859
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Might be related to #20438 somehow?

```
$ rustc --version
rustc 1.17.0 (56124baa9 2017-04-24)
```

### PR DESCRIPTION
**Pull request by euclio**
Fixes #41813.

The diagnostic looks something like this:

```
error: couldn't generate documentation: No space left on device (os error 28)
  |
  = note: failed to create or modify "/path/to/crate/target/doc/src/lazycell"
```

### CODE REVIEW

**Comment by TimNN** [comment]
Ping form triage @steveklabnik  / @rust-lang/rustdoc: This PR requires your review.

**Comment by steveklabnik** [comment]
@QuietMisdreavus or @GuillaumeGomez should be r+'ing this, not me

**Comment by GuillaumeGomez** [comment]
Looks nice! Please add a `rustdoc-ui` test and then I r+.

**Comment by euclio** [comment]
I'd like to, but I'm not sure how to simulate filesystem errors in a UI test.

**Comment by GuillaumeGomez** [comment]
Ah that's a good point...

@rust-lang/infra Does anyone knows how to simulate a full disk?

**Comment by kennytm** [comment]
what about write to a location which you have no permission? this will need to be changed to a run-make test though.

**Comment by GuillaumeGomez** [comment]
Hard to write an all-OSes test...

**Comment by QuietMisdreavus** [comment]
It may be enough to make it run on just Unix, though. The trick is just to make sure it outputs the right kind of error message when it encounters a filesystem error, even if the nature of that error doesn't matter much. Plus, plenty of other `run-make-fulldeps` tests skip running on Windows, so at least there's precedent there. At its core, we basically need to make sure that it doesn't say "internal compiler error" (or whatever the right capitalization should be) when it hits this error, right? If you want, i could try to sketch it out.

Otherwise, i quite like this PR. (I'm also not opposed to landing it as-is, since it's effectively just changing the way rustdoc prints certain kinds of errors.)

**Comment by euclio** [comment]
@QuietMisdreavus pushed a run-make test that checks that it doesn't ICE.

File: src/test/run-make-fulldeps/rustdoc-io-error/Makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# Ignore Windows: the test uses `chmod`.
+ifndef IS_WINDOWS
+
+# This test operates by creating a temporary directory and modifying its
+# permissions so that it is not writable. We have to take special care to set
+# the permissions back to normal so that it's able to be deleted later.
+all:
+	mkdir -p $(OUTPUT_DIR)
+	chmod u-w $(OUTPUT_DIR)
+	-$(shell $(RUSTDOC) -o $(OUTPUT_DIR) foo.rs)
```
Comment (QuietMisdreavus) [inline src/test/run-make-fulldeps/rustdoc-io-error/Makefile]: Does this save the exit status into `$(.SHELLSTATUS)` for later? I'm not familiar with this syntax.

Comment (euclio) [reply src/test/run-make-fulldeps/rustdoc-io-error/Makefile]: Yes. https://www.gnu.org/software/make/manual/html_node/Shell-Function.html#Shell-Function

File: src/librustdoc/lib.rs [hunk 2]
```diff
@@ -387,9 +388,24 @@ fn main_args(args: &[String]) -> isize {
...
+            krate,
+            renderopts,
+            passes.into_iter().collect(),
+            renderinfo,
+            &diag,
+        ) {
+            Ok(_) => rustc_driver::EXIT_SUCCESS,
+            Err(e) => {
+                diag.struct_span_err(
+                    DUMMY_SP,
```
[the remark this answers is not in the document]
Comment (euclio) [reply src/librustdoc/lib.rs]: Fixed, thanks!

**Review by QuietMisdreavus — APPROVED** [review]
Excellent, thanks so much! r=me when travis is green.

**Comment by kennytm** [comment]
@bors r- 

The test did not pass on Windows. See https://github.com/rust-lang/rust/pull/56531#issuecomment-444493202

```
---- [run-make] run-make-fulldeps\rustdoc-io-error stdout ----
error: make failed
status: exit code: 2
command: "make"
stdout:
------------------------------------------
make[1]: Entering directory '/c/projects/rust/src/test/run-make-fulldeps/rustdoc-io-error'
make[1]: Leaving directory '/c/projects/rust/src/test/run-make-fulldeps/rustdoc-io-error'
------------------------------------------
stderr:
------------------------------------------
make[1]: *** No targets.  Stop.
------------------------------------------
thread '[run-make] run-make-fulldeps\rustdoc-io-error' panicked at 'explicit panic', src\tools\compiletest\src\runtest.rs:3284:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failures:
    [run-make] run-make-fulldeps\rustdoc-io-error
test result: FAILED. 192 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
thread 'main' panicked at 'Some tests failed', src\tools\compiletest\src\main.rs:503:22
```

**Comment by euclio** [comment]
The test was missing an else branch that runs an empty target. Fixed.

**Comment by QuietMisdreavus** [comment]
r=me pending travis

**Comment by michaelwoerister** [comment]
I'm wondering why this caused a visible performance regression:
https://perf.rust-lang.org/compare.html?start=b866f7d258a2428e004039eb0f3fabd73cc9d6ae&end=14997d56a550f4aa99fe737593cd2758227afc56&stat=instructions:u

cc @rust-lang/wg-compiler-performance

**Comment by eddyb** [comment]
@michaelwoerister That makes no sense, I'd suspect an issue with the perf setup.

**Pull request merged**